### PR TITLE
Add Ice 2 Slice  translator

### PR DESCRIPTION
**Pull request by pepone**
This PR adds an `ice2slice` translator, that allows to convert Ice `.ice` files to IceRPC `.slice` files.

There are still a few things to do, especially to deal with things that don't have equivalent in `.slice` files.

### CODE REVIEW

File: cpp/src/ice2slice/Main.cpp [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+            if(parseStatus == EXIT_FAILURE)
+            {
+                status = EXIT_FAILURE;
+            }
+            else
+            {
+                DefinitionContextPtr dc = p->findDefinitionContext(p->topLevelFile());
+                assert(dc);
+                const string prefix = "js:module:";
```
Comment (bernardnormier) [inline cpp/src/ice2slice/Main.cpp:229]: Why this `js:module` processing?

File: cpp/src/ice2slice/Gen.cpp [hunk 2]
```diff
@@ -0,0 +1,721 @@
...
+        "uint8",                // KindByte
+        "bool",                 // KindBool
+        "int16",                // KindShort
+        "int32",                // KindInt
+        "int64",                // KindLong
+        "float32",              // KindFloat
+        "float64",              // KindDouble
+        "string",               // KindString
+        "???",                  // KindObject
+        "::Ice::ObjectProxy",   // KindObjectProxy
```
Comment (bernardnormier) [inline cpp/src/ice2slice/Gen.cpp:195]: I would map `Object*` to `::IceRpc::ServiceAddress?`.

File: cpp/src/ice2slice/Gen.cpp [hunk 3]
```diff
@@ -0,0 +1,721 @@
...
+    {
+        "uint8",                // KindByte
+        "bool",                 // KindBool
+        "int16",                // KindShort
+        "int32",                // KindInt
+        "int64",                // KindLong
+        "float32",              // KindFloat
+        "float64",              // KindDouble
+        "string",               // KindString
+        "???",                  // KindObject
```
Comment (bernardnormier) [inline cpp/src/ice2slice/Gen.cpp:194]: It's mapped to `AnyClass?`.

File: cpp/src/ice2slice/Gen.cpp [hunk 4]
```diff
@@ -0,0 +1,721 @@
...
+        "::Ice::ObjectProxy",   // KindObjectProxy
+        "???",                  // KindLocalObject
+        "???"                   // KindValue
```
Comment (bernardnormier) [inline cpp/src/ice2slice/Gen.cpp:197]: Also mapped to `AnyClass?`.

File: cpp/src/ice2slice/Gen.cpp [hunk 5]
```diff
@@ -0,0 +1,721 @@
...
+
+    if (optional)
+    {
+        os << "?";
+    }
+
+    return os.str();
+}
+
+string typeToCsString(const TypePtr& type, bool optional)
```
Comment (bernardnormier) [inline cpp/src/ice2slice/Gen.cpp:220]: Not clear why we need this in ice2slice.

**Review by bernardnormier — APPROVED** [review]

**Review by bernardnormier — APPROVED** [review]

File: cpp/src/ice2slice/Main.cpp (hunk 1 above)
Comment (pepone) [reply cpp/src/ice2slice/Main.cpp:229]: Need to remove, took slice2js as a template to start this project.

File: cpp/src/ice2slice/Gen.cpp (hunk 5 above)
Comment (pepone) [reply cpp/src/ice2slice/Gen.cpp:220]: To map cs::generic metadata to cs::type

File: cpp/src/ice2slice/Gen.cpp [hunk 6]
```diff
@@ -0,0 +1,721 @@
...
+    out << nl;
+    return false;
+}
+
+bool
+Gen::TypesVisitor::visitExceptionStart(const ExceptionPtr& p)
+{
+    if (p->isLocal())
+    {
+        // No loca types in Slice
```
Comment (externl) [inline cpp/src/ice2slice/Gen.cpp:541]:
```suggestion
        // No local types in IceRPC Slice
```

**Review by externl — APPROVED** [review]
Looks good!

**Review by InsertCreativityHere — APPROVED** [review]
Looks good!

**Comment by pepone** [comment]
I will deal with the comments in a follow-up PR.

**Pull request merged**